gpflow/utilities/utilities.py

```diff
@@ -246,22 +248,93 @@ def clear_bijector(bijector, _, state):
     return input_module
 
 
-M = TypeVar('M', bound=tf.Module)
+def _get_by_name_index(parent: object, attr_str: str, index_str: str) -> object:
+    attr = getattr(parent, attr_str)
+    if index_str != "":
+        try:
+            index = int(index_str)
+        except ValueError:
+            raise ValueError(f"Invalid index passed in the path'")
+        return attr[index]
+    return attr
+
+
+def _set_by_name_index(parent: object, value: Any, attr_str: str, index_str: str):
+    if index_str != "":
+        try:
+            index = int(index_str)
+        except ValueError:
+            raise ValueError(f"Invalid index passed in the path")
+        attr = getattr(parent, attr_str)
+        attr[index] = value
+    else:
+        setattr(parent, attr_str, value)
+
+
+def _get_last_attr_spec(parent: Any, attr_path: str) -> Any:
+    restr = r"((\w+)(\[(-?[\s\d]+)\])?(?<!\.))+"
```

Comment:
What does this do ?

Reply:
can you add your example you gave me this morning to the docstrings please